pkg/core/clock.go

```
@@ -0,0 +1,102 @@
+// limitations under the License.
+
+package core
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/time/utils"
+)
+
+type ClockOption struct {
+	CommonAttackConfig
+
+	Pid           int
+	SecDelta      int64
+	NsecDelta     int64
+	ClockIdsSlice string
+
+	CheckPidExist bool
+	Store         ClockFuncStore
+
+	ClockIdsMask uint64
+}
+
+type ClockFuncStore struct {
+	CodeOfGetClockFunc []byte
+	OriginAddress      uint64
+}
+
+func NewClockOption() *ClockOption {
+	return &ClockOption{
+		CommonAttackConfig: CommonAttackConfig{
+			Kind: ClockAttack,
+		},
+	}
+}
+
+func (opt *ClockOption) PreProcess() error {
+	clkIds := strings.Split(opt.ClockIdsSlice, ",")
+
+	clockIdsMask, err := utils.EncodeClkIds(clkIds)
+	if err != nil {
+		log.Error("error while converting clock ids to mask", zap.Error(err))
+		return err
+	}
+	if clockIdsMask == 0 {
+		log.Error("clock ids must not be empty")
+		return fmt.Errorf("clock ids must not be empty")
+	}
+	opt.ClockIdsMask = clockIdsMask
+
+	if uint64(opt.SecDelta) > 1<<31 {
+		log.Warn("Monotonic clock will be broken when sec delta is too large or too small.")
+		if uint64(opt.SecDelta) > 1<<56 {
+			log.Warn("Time zone info will be broken when sec delta is too large or too small.")
+		}
+	}
+
+	if uint64(opt.NsecDelta) > 1<<56 {
+		log.Warn("Time will be broken when nanosecond delta is too large or too small")
+	}
+
+	if opt.CheckPidExist {
+		// Since os.FindProcess in unix systems will always succeed
+		// regardless of whether the process exists (https://pkg.go.dev/os#FindProcess),
+		// we need to use process.Signal to check if pid is accessible.
+		process, err := os.FindProcess(opt.Pid)
+		if err != nil {
+			fmt.Printf("Failed to find process: %s\n", err)
+		} else {
+			err := process.Signal(syscall.Signal(0))
+			if err != nil {
+				fmt.Printf("Pid may not be accessible , because : %v", err)
```

is it need to return this error?

---

Fixed.